takeshape/penny
# Improve branches workflow

This PR updates the branch workflow to:

- Create branches based on a git hook
- Promote branches based on a hook (locally) or during a build in CI
- Use your branch URL everywhere you would typically use an API URL
- Better integrate branches into the whole process
- Supports non-interactive create / merge if using VSCode or other git tools that don't have TTY

### Test Plan (Steps to test):

#### Branch Creation / PR branch use

1. Pull this branch
1. Add a `.env.local` file with a `TAKESHAPE_API_KEY='{API KEY WITH BRANCH PERMS}'`
1. `npm install`
1. Create a new branch off this branch, `git checkout -b my-new-branch-RANDOM`
3. You should be prompted to create a new API branch, press enter (yes)
4. The branch should be created. Check your client to see if it is there.
5. Push your branch and open a PR
6. When Vercel builds, check it and see if it uses a versioned URL from your branch (I typically sign in to get a live request and filter for `graphql`).
7. When Netlify builds, do the same
8. Close the PR
9. Delete the branch

#### Branch Promotion in CI

1. Pull the latest `main` branch
2. Create a new branch, `my-new-branch-RANDOM`
3. In the client, create a new branch, `my-new-branch-RANDOM`.
4. Create a shape in your API branch, `MyNewShapeRandom` or similar
5. Make some minor change to your git branch, append a dot to the README.md or something.
6. Push your git branch and open a PR for it
7. Let it run for a minute or so, then merge the PR
8. Check Vercel, review the build log for `main` / `production`. You should see a note about `my-new-branch-RANDOM` being promoted and deleted near the top.
10. Check the client, verify the branch is deleted and that your shape made it into production.
11. Do this all with Netlify as well

#### Branch Promotion in git

1. `git checkout takeshape/merge-test`
2. Add a `.env.local` file with a `TAKESHAPE_API_KEY='{API KEY WITH BRANCH PERMS}'`
3. `npm install`
4. `git checkout -b my_test`
5. `echo "." >> README.md`
6. `git commit -am "update readme"`
8. `git checkout main`
9. `git merge my_test`
10.  ... prompt, Y
11. Merging... merged and deleted
12. Check the client, was your branch deleted?

#### Shape commands

1. Pull this branch
2. Run `npm run takeshape:create-branch -- --name my_test_branch`.
3. Verify the branch was created in the client
4. Run `npm run takeshape:delete-branch -- --name my_test_branch`
5. Verify the branch was deleted in the client
6. Run `npm run takeshape:create-branch -- --name my_test_branch` again
7. Run `npm run takeshape:promote-branch -- --name my_test_branch`
8. Verify the branch was deleted

#### Review the lib

1. Check out the code over here: https://github.com/takeshape/shape-tools which does all the heavy lifting

### Checklist:
